velox/exec/Window.cpp

```diff
@@ -137,12 +137,49 @@ void Window::createWindowFunctions(
         operatorCtx_->pool(),
         &stringAllocator_));
 
-    windowFrames_.push_back(
-        {windowNodeFunction.frame.type,
-         windowNodeFunction.frame.startType,
-         windowNodeFunction.frame.endType,
-         fieldArgToChannel(windowNodeFunction.frame.startValue),
-         fieldArgToChannel(windowNodeFunction.frame.endValue)});
+    auto createFrameChannelArg =
+        [&](const core::TypedExprPtr& frame) -> FrameChannelArg {
+      if (auto constant = constantArg(frame)) {
+        auto constantOffset =
+            constant->template as<ConstantVector<int64_t>>()->valueAt(0);
+        VELOX_USER_CHECK_GE(
+            constantOffset, 1, "k in frame bounds must be at least 1");
+        return {kConstantChannel, constant};
+      } else {
+        return {
+            fieldArgToChannel(frame).value(),
+            BaseVector::create(BIGINT(), 0, operatorCtx_->pool())};
+      }
+    };
+
+    bool hasKBasedStartBound =
+        (windowNodeFunction.frame.startType ==
+             core::WindowNode::BoundType::kPreceding ||
+         windowNodeFunction.frame.startType ==
+             core::WindowNode::BoundType::kFollowing);
+    bool hasKBasedEndBound =
+        (windowNodeFunction.frame.endType ==
+             core::WindowNode::BoundType::kPreceding ||
+         windowNodeFunction.frame.endType ==
+             core::WindowNode::BoundType::kFollowing);
+    if (windowNodeFunction.frame.type == core::WindowNode::WindowType::kRange &&
+        (hasKBasedStartBound || hasKBasedEndBound)) {
+      VELOX_NYI("k preceding/following is only supported in ROWS mode");
+    }
+
+    WindowFrame windowFrame({
```

Comment:
You can create this in-place in push_back call below (like the original code in the lhs).

Reply:
Updated, thanks!